module/zfs/dmu_send.c

```diff
@@ -587,15 +597,49 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
 		}
 	}
 
+	/*
+	 * The DRR_SPILL_BLOCK flag is set for every dnode which references
+	 * a spill block.  This allows the receiving pool to definitively
+	 * determine when a spill block should be kept or freed.
+	 */
+	if (dsp->dsa_spill && (dnp->dn_flags & DNODE_FLAG_SPILL_BLKPTR))
+		drro->drr_flags |= DRR_SPILL_BLOCK;
+
 	if (dump_record(dsp, DN_BONUS(dnp), bonuslen) != 0)
 		return (SET_ERROR(EINTR));
 
 	/* Free anything past the end of the file. */
 	if (dump_free(dsp, object, (dnp->dn_maxblkid + 1) *
 	    (dnp->dn_datablkszsec << SPA_MINBLOCKSHIFT), DMU_OBJECT_END) != 0)
 		return (SET_ERROR(EINTR));
+
+	/*
+	 * Send extra DRR_SPILL records.  By default, these are included in
+	 * the stream in order to recreate spill blocks that are incorrectly
+	 * removed.  This may occur in previous versions of ZFS when certain
```

Comment:
What is the difference in correctness (if any) between receiving a non-raw stream (generated by the new code with the default tunables) on old software vs on new software (with this change)?

Reply:
With this change both old and new software receiving the non-raw stream will not have any differences in correctness.  The one caveat to this, is that because we're always sending the spill block additional versions of the exact same spill block will be created on the old software.  The new software will correctly preserve existing spill blocks and discard the extra DRR_SPILL records.